modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java

```diff
@@ -1361,18 +1361,74 @@ public static void setSubsamplingFactors(
         // Note: using "round" instead of floor to go for the closest subsampling factory
         // improves quality
         // /////////////////////////////////////////////////////////////////////
-        int subSamplingFactorX = (int) Math.round(requestedRes[0] / selectedRes[0]);
-        subSamplingFactorX = subSamplingFactorX == 0 ? 1 : subSamplingFactorX;
 
-        while (subSamplingFactorX > 0 && rasterWidth / subSamplingFactorX <= 0) subSamplingFactorX--;
-        subSamplingFactorX = subSamplingFactorX <= 0 ? 1 : subSamplingFactorX;
-
-        int subSamplingFactorY = (int) Math.round(requestedRes[1] / selectedRes[1]);
-        subSamplingFactorY = subSamplingFactorY == 0 ? 1 : subSamplingFactorY;
+        // Control parameters
+        final double resEpsilon = 1e-9; // Small epsilon to avoid floating point errors when using floor
+        final double overviewTolerance =
+                0.03; // Relative tolerance (%) to allow a small upward adjustment of subsampling
+
+        // Ratios between requested resolution and the resolution of the selected overview
+        final double ratioX = requestedRes[0] / selectedRes[0];
+        final double ratioY = requestedRes[1] / selectedRes[1];
+
+        // Use floor with epsilon and minimum clamps
+        int subSamplingFactorX = (int) Math.floor(ratioX + resEpsilon);
+        int subSamplingFactorY = (int) Math.floor(ratioY + resEpsilon);
+
+        subSamplingFactorX = Math.max(1, subSamplingFactorX);
+        subSamplingFactorY = Math.max(1, subSamplingFactorY);
+
+        // Explicit clamp per dimension (replaces integer division)
+        // Ensures we never request a factor larger than the raster size
+        if (subSamplingFactorX > rasterWidth) subSamplingFactorX = rasterWidth;
+        if (subSamplingFactorY > rasterHeight) subSamplingFactorY = rasterHeight;
+
+        // Upward adjustment ONLY if it falls within tolerance
```

Comment:
This needs to be explained better. Something like:

> The previous code used floor to compute an integer subsampling factor. Now testing if switching to the next integer (ceil) is within tolerance, and in that case, use it to get better performance. 

(and was it for better performance? I think so, reading less pixels, but want to double check)